test/token_rules/register_rule.js

```diff
@@ -0,0 +1,283 @@
+
+            await utils.expectRevert(
+                tokenRules.registerRule(
+                    ruleName,
+                    ruleAddress,
+                    ruleAbi,
+                    {
+                        from: organization,
+                    },
+                ),
+                'Rule with an empty ABI can not be registered.',
+            );
+        });
+    });
+
+    contract('Events', async (accounts) => {
+        it('RuleRegistered is emitted.', async () => {
+            const organization = accounts[0];
+            const token = accounts[1];
+
+            const aRuleName = 'A';
+            const aRuleAddress = accounts[2];
+            const aRuleAbi = `Rule abi of ${aRuleName}`;
+
+            const tokenRules = await TokenRules.new(organization, token);
+
+            const transactionResponse = await tokenRules.registerRule(
+                aRuleName,
+                aRuleAddress,
+                aRuleAbi,
+                {
+                    from: organization,
+                },
+            );
+
+            const events = Event.decodeTransactionResponse(
+                transactionResponse,
+            );
+
+            assert.strictEqual(
+                events.length,
+                1,
+            );
+
+            Event.assertEqual(events[0], {
+                name: 'RuleRegistered',
+                args: {
+                    _ruleName: aRuleName,
+                    _ruleAddress: aRuleAddress,
+                },
+                logIndex: 0,
+            });
+        });
+    });
+
+    contract('Storage', async (accounts) => {
+        it('Registered rule exists in storage.', async () => {
+            const organization = accounts[0];
+            const token = accounts[1];
+
+            const aRuleName = 'A';
+            const aRuleAddress = accounts[2];
+            const aRuleAbi = `Rule abi of ${aRuleName}`;
+
+            const tokenRules = await TokenRules.new(organization, token);
+
+            await tokenRules.registerRule(
+                aRuleName,
+                aRuleAddress,
+                aRuleAbi,
+                {
+                    from: organization,
+                },
+            );
+
+            const aRuleIndex = 0;
+
+            const rule = await tokenRules.rules.call(aRuleIndex);
+
+            assert.strictEqual(
+                rule.ruleName,
+                aRuleName,
+            );
+
+            assert.strictEqual(
+                rule.ruleAddress,
+                aRuleAddress,
+            );
+
+            assert.strictEqual(
+                rule.ruleAbi,
+                aRuleAbi,
+            );
+
+            assert.isOk(
+                await tokenRules.rulesByNameHash.call(
+                    web3.utils.keccak256(aRuleName),
+                ),
+            );
+
+            assert.isOk(
+                await tokenRules.rulesByAddress.call(aRuleAddress),
+            );
+        });
+    });
+});
```

Comment:
Just to know we will continue write unit tests for other TokenRules methods like addGlobalConstraint, removeGlobalConstraint etc. It will be different PR.

Reply:
Sure.